### ISSUE TITLE
flutter pub upgrade fails with permissions error

### ISSUE THREAD

**Issue by robinbakkerus** [flutter/flutter#40331]
On my Windows 10 machine I constantly get an error while trying to an upgrade:

```
> flutter pub get
Failed to rename directory because access was denied.
This may be caused by a virus scanner or having a file
in the directory open in another application.
```
**Unfortunately it does not tell what directory can not be renamed. Is there a way to find out ?**

I have the latest flutter:

```
> flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.9.1+hotfix.2, on Microsoft Windows [Version 10.0.17134.950], locale nl-NL)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[!] Android Studio (version 3.4)
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
[√] Connected device (1 available)

```
I have McAffee running which I cannot stop and this company Dell laptop is also configured with automatic harddisk encryption.

**Comment by mingoal** [comment]
I have same issue. Unfortunately, can't turn off Macfee. Stucking...

**Comment by markballenger** [comment]
same @mingoal , i also have mcafee aka suckafee

**Comment by mingoal** [comment]
if you have this issue, try this. 
1. add `-v` for installing
`flutter pub global activate devtools -v`, it will print out something like
>  Failed to rename directory because access was denied. Retrying in 50ms.
IO  : Renaming directory C:\XXXXXX\bin\flutter\.pub-cache\_temp\dir34844307-29ce-11ea-9eb5-e86a648f9e9a to C:\XXXXXX\bin\flutter\.pub-cache\hosted\pub.flutter-io.cn\devtools-0.1.13
2. manually copy temp file to target directory
3. check it with `flutter pub global activate devtools`
> Precompiling executables...
Precompiled devtools:bin\devtools.
Installed executable devtools.
Activated devtools 0.1.13.

Not sure it works or not as vscode keep on saying need to install devtools.
**Good luck**

**Comment by regalme** [comment]
Been dealing with this for hours upon hours. 
Finally deleted .dart_tool from app's directory... ran flutter pub get again and it now works fine

**Comment by ratnapriya4g** [comment]
Building flutter tool...
Running pub upgrade...
Failed to rename directory because access was denied.
This may be caused by a virus scanner or having a file
in the directory open in another application.
Error (1): Unable to 'pub upgrade' flutter tool.

I am facing this problem again and again while upgrade the flutter

**Comment by CapTen101** [comment]
Yup, exactly the same here. @ratnapriya4g 
All environment variables are correctly set up.
Yet no solution for this issue yet.
PS. I'm on windows and I run Quick Heal

**Comment by pinhassi** [comment]
I run this:
`flutter pub global activate devtools -v`

To see the error.
**The problem was with the AntiVirus** (BitDefender in my case) preventing access to the cache folder.

**Comment by daniel-sc** [comment]
For me setting the environment variable `$PUB_CACHE` to some folder that was not checked by anti virus solved this.

**Comment by lightwaxx** [comment]
I was having this issue when I ran `flutter upgrade` and the cause was my Antivirus (Kaspersky). Just disable protection and it should complete the upgrade then enable protection.

**Comment by basemnoori1990** [comment]
if you have avast just uninstall it

**Comment by ratnapriya4g** [comment]
@CapTen101 
Firstly, **open your flutter console** and **run this**:
 `flutter pub global activate devtools`  then it will come like this:
![flutterPub](https://user-images.githubusercontent.com/38585678/81500558-84709b00-92f0-11ea-8965-adc7da0eca97.jpeg)

After running it completely, then you **run this:**
`flutter upgrade`, it will come like this:

![flutterUpgrade](https://user-images.githubusercontent.com/38585678/81500572-a4a05a00-92f0-11ea-864b-d938a27ae4f3.jpeg)
it will definitely work because I try it exactly, it's working fine.

**Comment by CapTen101** [comment]
@ratnapriya4g  Though I already figured it out that my Antivirus was causing the problem and excluding my flutter source from the scanning did the job for me, still Thanks a lot! 😄

**Comment by jmagman** [comment]
This is a request to improve https://github.com/dart-lang/pub/blob/03b7f9a5a27f46bf554f2a3bdef965e7354c69b0/lib/src/io.dart#L404 reason to include the directory path.

See https://github.com/dart-lang/pub/blob/03b7f9a5a27f46bf554f2a3bdef965e7354c69b0/lib/src/io.dart#L359.

Closing as duplicate of https://github.com/dart-lang/pub/issues/1798.  If you think that doesn't cover it, please file a new issue with [pub](https://github.com/dart-lang/pub/issues/new/choose).

**Comment by vinc3nati** [comment]
> 
> 
> if you have this issue, try this.
> 
>     1. add `-v` for installing
>        `flutter pub global activate devtools -v`, it will print out something like
> 
> 
> > Failed to rename directory because access was denied. Retrying in 50ms.
> > IO  : Renaming directory C:\XXXXXX\bin\flutter.pub-cache_temp\dir34844307-29ce-11ea-9eb5-e86a648f9e9a to C:\XXXXXX\bin\flutter.pub-cache\hosted\pub.flutter-io.cn\devtools-0.1.13
> 
>     1. manually copy temp file to target directory
> 
>     2. check it with `flutter pub global activate devtools`
> 
> 
> > Precompiling executables...
> > Precompiled devtools:bin\devtools.
> > Installed executable devtools.
> > Activated devtools 0.1.13.
> 
> Not sure it works or not as vscode keep on saying need to install devtools.
> **Good luck**

Hey which is the target directory?

**Comment by JohannaShiri** [comment]
I got the error, can't disable the antivirus. 
Spent hours solving this. 

was trying to get the "epub" package so... 

What worked for me: 

1. I opened the folder which contains all the packages (windows) 
C:\flutter\.pub-cache\hosted\pub.dartlang.org
2. "copy - paste" the relevant folder (in my case "epub-2.1.0"). If the folder isn't there already,  you can download it manually from github. 
3. Delete the original package folder and rename the copied folder to have the same name as the original one. 
so for me, I changed the name of the folder from: "epub-2.1.0 - Copy" to "epub-2.1.0"

running "flutter pub get" and had no errors.

**Comment by Atharva-Vijay-Khade** [comment]
> @ratnapriya4g Though I already figured it out that my Antivirus was causing the problem and excluding my flutter source from the scanning did the job for me, still Thanks a lot! 😄

i am facing the same issue, please could you tell that how to exclude the flutter source from the scanning?

**Comment by silvioascruz** [comment]
I get same error. I see that occurs when execute the follow command in c:\src\flutter_beta\bin\cache\dart-sdk\bin\pub.bat:
`if defined USING_DART_1 (
  echo "Pub no longer supports Dart 1"
) else (
  "%BIN_DIR%\dart" %VM_OPTIONS% "%BIN_DIR%\snapshots\pub.dart.snapshot" %*
)
`
I insert a comment after and before command, to ensure that local:
`if defined USING_DART_1 (
  echo "Pub no longer supports Dart 1"
) else (
  ECHO "-----------------------------------------------------"
   ECHO "Command: "%BIN_DIR%\dart" %VM_OPTIONS% "%BIN_DIR%\snapshots\pub.dart.snapshot" %*
  "%BIN_DIR%\dart" %VM_OPTIONS% "%BIN_DIR%\snapshots\pub.dart.snapshot" %*
  ECHO "-----------------------------------------------------"
)
`
The  result, after execute any flutter command is:


PS C:\> flutter pub get
`
Building flutter tool...
Running pub upgrade...
`
`
"-----------------------------------------------------"
`

`
"Command: "C:\src\flutter_beta\bin\cache\dart-sdk\bin\dart"  "C:\src\flutter_beta\bin\cache\dart-sdk\bin\snapshots\pub.dart.snapshot" upgrade "--verbosity=error" --no-precompile
`

`
Pub failed to rename directory because access was denied.
This may be caused by a virus scanner or having a file in the directory open in another application.
`
`
"-----------------------------------------------------"
`

`
Error (1): Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries left)
`

**Comment by apzentral** [comment]
In my case, this has been blocked by AVG Behavior Shield. Once I turn this off, it now work.

**Comment by loperdev** [comment]
I had the same issue, so after trying all the methods provides above I had the same problem so I downloaded flutter again and replaced it with the current flutter folder. It solved my problem.

**Comment by jessechen0319** [comment]
> > if you have this issue, try this.
> > ```
> > 1. add `-v` for installing
> >    `flutter pub global activate devtools -v`, it will print out something like
> > ```
> > 
> > 
> > > Failed to rename directory because access was denied. Retrying in 50ms.
> > > IO  : Renaming directory C:\XXXXXX\bin\flutter.pub-cache_temp\dir34844307-29ce-11ea-9eb5-e86a648f9e9a to C:\XXXXXX\bin\flutter.pub-cache\hosted\pub.flutter-io.cn\devtools-0.1.13
> > 
> > 
> > ```
> > 1. manually copy temp file to target directory
> > 
> > 2. check it with `flutter pub global activate devtools`
> > ```
> > 
> > 
> > > Precompiling executables...
> > > Precompiled devtools:bin\devtools.
> > > Installed executable devtools.
> > > Activated devtools 0.1.13.
> > 
> > 
> > Not sure it works or not as vscode keep on saying need to install devtools.
> > **Good luck**
> 
> Hey which is the target directory?

Done, working now
